python/lsst/cp/pipe/defects.py

```diff
@@ -0,0 +1,645 @@
+
+        elif self.config.mode == 'VISITS':
+            butler = dataRef.getButler()
+
+            if self.config.assertSameRun:
+                runs = self._getRunListFromVisits(butler, visitList)
+                if len(runs) != 1:
+                    raise RuntimeError(f'Got data from runs {runs} with assertSameRun==True')
+
+            for visit in visitList:
+                imageType = butler.queryMetadata('raw', self.config.imageTypeKey, dataId={'visit': visit})[0]
+                imageType = imageType.lower()
+                dataRef.dataId['visit'] = visit
+
+                if imageType == 'flat':  # note different isr tasks
+                    exp = self.isrForFlats.runDataRef(dataRef).exposure
+                    defects = self.findHotAndColdPixels(exp, imageType)
+                    defectListFromFlats.append(defects)
+
+                elif imageType == 'dark':
+                    exp = self.isrForDarks.runDataRef(dataRef).exposure
+                    defects = self.findHotAndColdPixels(exp, imageType)
+                    defectListFromDarks.append(defects)
+
+                else:
+                    raise RuntimeError("Failed on imageType {imageType}. Only flats and darks supported")
+
+                msg = f"Found {len(defects)} defects containing {self._nPixFromDefects(defects)} pixels"
+                msg += f" in visit {visit}"
+                self.log.info(msg)
+
+                if self.config.makePlots:
+                    nSig = self.config.nSigmaBright if imageType == 'flat' else self.config.nSigmaDark
+                    self._plot(dataRef, exp, visit, nSig, defects, imageType)
+
+        self.log.info(f'Combining {len(defectListFromDarks)} defect sets from darks for detector {detNum}')
+        mergedDefectsFromDarks = self._postProcessDefectSets(defectListFromDarks, exp.getDimensions())
+        self.log.info(f'Combining {len(defectListFromFlats)} defect sets from flats for detector {detNum}')
+        mergedDefectsFromFlats = self._postProcessDefectSets(defectListFromFlats, exp.getDimensions())
+
+        self.log.info(f'Combining bright and dark defect sets')
+        brightDarkPostMerge = [mergedDefectsFromDarks, mergedDefectsFromFlats]
+        allDefects = self._postProcessDefectSets(brightDarkPostMerge, exp.getDimensions(), mode='OR')
+
+        self._writeData(dataRef, allDefects)
+
+        self.log.info('Finished finding defects in detector %s' % detNum)
+        return pipeBase.Struct(exitStatus=0)
+
+    @staticmethod
+    def _nPixFromDefects(defects):
+        """TODO: Write docstring
+        """
+        nPix = 0
+        for defect in defects:
+            nPix += defect.getBBox().getArea()
+        return nPix
+
+    def _writeData(self, dataRef, defects):
+        """Write the data out to the defect file.
+
+        TODO: Add parameters
+        """
+        filename = dataRef.getUri(write=True)  # does not guarantee that full path exists
```

Comment:
Yes, because the format is being decided by the config option, and also I don't think the butler knows how to put defects generically. I realise this is not ideal 😕

Reply:
I'm not clear why butler can't handle a `lsst.meas.algorithms.Defects` object directly. Now that defects are real classes rather than generic `list` this is all doable.   @SimonKrughoff spent a lot of work on this. Is the issue that you want the butler to put them in a special calibration repository?

Reply:
Oh, so your real problem is not butler handling of `Defects`, it's that you can't tell the butler whether to write a text file or fits file at `put` time? That's a butler configuration option not a task configuration option. Are you sure it shouldn't be a butler configuration option?

Reply:
I'm not even sure what that means. If I'm running the task, and I want the output to be a human readable text file, how do I do that?

Reply:
I know how to do it in Gen3 (you would configure a formatter for your storage class that wrote it in the format you want) although not as a run time configuration.  Gen2 butler is I think set up to write a FITS file by default. A different formatter would be needed for the text file and again I'm not sure how to change that behavior at run time. cc/ @ktlim ?

Reply:
Is this worth caring about? (and also note that currently there is an option of "BOTH" too...)

Reply:
I don't even know if anyone really cares about this functionality tbh, I just threw it in there without much thought. I'd just like to do something with minimal resistance really, so just tell me what to do and I'll do it.

Reply:
Leave it for now. For Gen3 we will need to ponder how to handle it.

Reply:
🙏